app/flatpak-builtins-alias.c

```diff
@@ -0,0 +1,233 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthew Leeds <mwleeds@protonmail.com>
+ */
+
+#include "config.h"
+
+#include <locale.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#include <glib/gi18n.h>
+
+#include "libglnx.h"
+
+#include "flatpak-builtins.h"
+#include "flatpak-builtins-utils.h"
+#include "flatpak-cli-transaction.h"
+#include "flatpak-quiet-transaction.h"
+#include "flatpak-utils-private.h"
+#include "flatpak-error.h"
+#include "flatpak-table-printer.h"
+
+static gboolean opt_remove;
+
+static GOptionEntry options[] = {
+  { "remove", 0, 0, G_OPTION_ARG_NONE, &opt_remove, N_("Remove the specified alias"), NULL },
+  { NULL }
+};
+
+gboolean
+flatpak_builtin_alias (int argc, char **argv, GCancellable *cancellable, GError **error)
+{
+  g_autoptr(GOptionContext) context = NULL;
+  g_autoptr(GPtrArray) dirs = NULL;
+  int i;
+
+  context = g_option_context_new (_("ALIAS [APP] - Add an alias for running the app APP"));
+  g_option_context_set_translation_domain (context, GETTEXT_PACKAGE);
+
+  if (!flatpak_option_context_parse (context, options, &argc, &argv,
+                                     FLATPAK_BUILTIN_FLAG_ALL_DIRS | FLATPAK_BUILTIN_FLAG_OPTIONAL_REPO,
+                                     &dirs, cancellable, error))
+    return FALSE;
+
+  /* Move the user dir to the front so it "wins" in case an app is in more than
+   * one installation */
+  if (dirs->len > 1)
+    {
+      /* Walk through the array backwards so we can safely remove */
+      for (i = dirs->len; i > 0; i--)
+        {
+          FlatpakDir *dir = g_ptr_array_index (dirs, i - 1);
+          if (flatpak_dir_is_user (dir))
+            {
+              g_ptr_array_insert (dirs, 0, g_object_ref (dir));
+              g_ptr_array_remove_index (dirs, i);
+              break;
+            }
+        }
+    }
+
+  if (argc == 1)
+    {
+      g_autoptr(FlatpakTablePrinter) printer = NULL;
+
+      printer = flatpak_table_printer_new ();
+      flatpak_table_printer_set_column_title (printer, 0, _("Alias"));
+      flatpak_table_printer_set_column_title (printer, 1, _("App"));
+      flatpak_table_printer_set_column_title (printer, 2, _("Installation"));
+
+      for (i = 0; i < dirs->len; i++)
+        {
+          FlatpakDir *dir = g_ptr_array_index (dirs, i);
+          g_autoptr(GHashTable) aliases = NULL; /* alias → app-id */
+
+          aliases = flatpak_dir_get_aliases (dir);
+
+          GLNX_HASH_TABLE_FOREACH_KV (aliases, const char *, alias, const char *, app_id)
```

Comment:
This seems a bit weird. It should loop over all aliases in all dirs, with the first one "winning" if there are duplicates. Then it should create the table, so as to avoid duplicates. We should also flatpak_table_printer_set_key() on the alias and sort on that.